### DISCUSSION TITLE
What NONITERACTIVE stands for?

### DISCUSSION THREAD

**Discussion by guandalf** [Homebrew/discussions#4547] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.20-2-ga380aed
ORIGIN: https://github.com/Homebrew/brew
HEAD: a380aed887f40cbe29953269798d510502bf177d
Last commit: 23 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 115de939f93ff95e858a759eec59da456fe89343
Core tap last commit: 4 months ago
Core tap branch: master
Core tap JSON: 31 May 14:45 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.XbHVkfhXds/org.xquartz:0
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit westmere
Clang: 14.0.3 build 1403
Git: 2.40.1 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-x86_64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: true
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  pycparser
  cffi

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  go@1.17
  librem
  node@14
```

### Description of issue

I really cannot understand the purpose of the noninteractive if neither running it as normal user nor running it as sudo works :) 

~ ⌚ 17:05:15
$ NONINTERACTIVE=1 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Running in non-interactive mode because `$NONINTERACTIVE` is set.
==> Checking for `sudo` access (which may request your password)...
Need sudo access on macOS (e.g. the user guandalf needs to be an Administrator)!

~ ⌚ 17:05:22
$ sudo NONINTERACTIVE=1 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
Password:
==> Running in non-interactive mode because `$NONINTERACTIVE` is set.
==> Checking for `sudo` access (which may request your password)...
Don't run this as root!

~ ⌚ 17:06:09
$

**Comment by gromgit** [comment, marked as answer]
Setting the `NONINTERACTIVE` environment variable tells the Homebrew _installer_ to not interact with the user at all. At this point, this is basically restricted to running `sudo` in "do not prompt for password" mode, which will fail if:
* `sudo` wasn't configured to let you execute any command as root without a password, and
* you haven't successfully invoked `sudo` _with_ a password in the last 15 minutes or so (or however long `sudo` is configured to cache your credentials after successful authentication)

Your first run failed because of the above conditions. If you want to use `NONINTERACTIVE`, you either have to add something like the following to `/etc/sudoers`:
```
# Let guandalf OR ANYONE WHO HACKS HIS ACCOUNT execute EVERYTHING as root without authentication
guandalf   ALL=NOPASSWD: ALL
```
or run `sudo -v` to "preauthenticate" yourself to `sudo` before running `NONINTERACTIVE=1 /bin/bash ...`

Your second run failed because the Homebrew installer absolutely doesn't want to do _everything_ as root. Down that road lies madness.
